File: sample_code/read.cpp
#include "read.h"

#include <iostream>
#include <fstream>
using namespace std;

bool readfile(string filename, Plan& plan, VariableTracker& tracker) {
	ifstream infile(filename);
	if (!infile) {
		cerr << "Could not read file.\n";
		return false;
	}

	string identifier;
	int locations, robots, cranes, piles, containers;
	infile >> identifier >> locations;
	infile >> identifier >> robots;
	infile >> identifier >> cranes;
	infile >> identifier >> piles;
	infile >> identifier >> containers;

	tracker = VariableTracker(locations, robots, cranes, piles, containers);
	plan.nextVar = tracker.getFirstVar();

	Action start(START);
	Action end(FINISH);
	bool goal = false;
	while (infile) {
		infile >> ws; //Get rid of whitespace first?
		if (infile.peek() == '#') {
			getline(infile, identifier, '\n');
			continue;
		}
		infile >> identifier;
		if (!infile) continue;
		if (identifier == "initial") continue;
		if (identifier == "goal") {
			goal = true;
			continue;
		}
		Predicates id;
		try {
			id = (Predicates) Name2Predicate.at(identifier);
		}
		catch (out_of_range e) {
			cerr << "Bad identifier\n";
			return false;
		}
		Predicate newPred;
		string arg1, arg2;
		switch (id) {
			// Two-argument propositions
			case ADJACENT:
			case ATTACHED:
			case BELONG:
			case AT:
			case LOADED:
			case HOLDING:
			case IN:
			case ON:
			case TOP:
				infile >> arg1 >> arg2;
				if (!infile) {
					cerr << "Incomplete predicate\n";
					return false;
				}
				newPred = Predicate(id, tracker.getId(arg1), tracker.getId(arg2));
			break;
			// One-argument propositions
			case OCCUPIED:
			case UNLOADED:
			case EMPTY:
			case FREE:
				infile >> arg1;
				if (!infile) {
					cerr << "Incomplete predicate\n";
					return false;
				}
				newPred = Predicate(id, tracker.getId(arg1));
			break;
		};
		if (goal) plan.open.push_back(make_pair(newPred,1));
		else start.addList.push_back(newPred);
	}
	plan.steps.push_back(start);
	plan.steps.push_back(end);
	//plan.orderings.push_back(make_pair(0,1)); // start < finish
	plan.orderings.insert(Ordering(0,1));

	return true;
}

void printPredicate(ostream& out, const Predicate& p, 
	const VariableTracker& tracker) 
{
	out << Predicate2Name.at(p.type);
	for (int i = 0; i < 2; ++i) {
		if (p.args[i] < 0) break;
		out << ' ' << tracker.getName(p.args[i]);
	}
}

void printAction(ostream& out, const Action& A, 
	const VariableTracker& tracker) 
{
	out << Action2Name.at(A.type);
	for (int i = 0; i < 5; ++i) {
		if (A.args[i] < 0) break;
		out << ' ' << tracker.getName(A.args[i]);
	}
	out << '\n';
	out << "     Prereqs: ";
	vector<Predicate> prereqs = A.getPrereqs();
	for (int i = 0; i < prereqs.size(); ++i) {
		printPredicate(out, prereqs[i], tracker);
		out << ", ";
	}
	out << '\n';
	out << "     Added: ";
	for (int i = 0; i < A.addList.size(); ++i) {
		printPredicate(out, A.addList[i], tracker);
		out << ", ";
	}
	out << '\n';
	out << "     Deleted: ";
	for (int i = 0; i < A.deleteList.size(); ++i) {
		printPredicate(out, A.deleteList[i], tracker);
		out << ", ";
	}
	out << '\n';
}

void printPlan(string outfilename, const Plan& plan, const VariableTracker& tracker) {
	ofstream outfile(outfilename);
	if (outfile) {
		printPlan(outfile, plan, tracker);
	}
	else {
		cerr << "Could open output file.\n";
	}
}

void printPlan(ostream& out, const Plan& plan, const VariableTracker& tracker) {
	out << "actions\n";
	int startStep = -1, endStep = -1;
	for (int i = 0; i < plan.steps.size(); ++i) {
		const Action& act = plan.steps[i];
		if (act.type == START) startStep = i;
		if (act.type == FINISH) endStep = i;
		out << i << ' ' << Action2Name.at(act.type);
		for (int j = 0; j < 5; ++j) {
			if (act.args[j] < 0) break;
			out << ' ' << tracker.getName(act.args[j]);
		}
		out << '\n';
	}
	out << "\nconstraints\n";
	for (auto i = plan.orderings.begin(); i != plan.orderings.end(); ++i) {
		const Ordering& o = *i;
		// Skip over start / end ordering constraints
		if (o.first == startStep || o.second == endStep) continue;
		out << o.first << " < " << o.second << '\n';
	}
}

void printVerbosePlan(ostream& out, const Plan& plan, const VariableTracker& tracker) {
	out << "#Steps\n";
	for (int i = 0; i < plan.steps.size(); ++i) {
		const Action& act = plan.steps[i];
		out << i << ' ' << Action2Name.at(act.type);
		for (int j = 0; j < 5; ++j) {
			if (act.args[j] < 0) break;
			out << ' ' << tracker.getName(act.args[j]);
		}
		out << '\n';
	}
	out << "\n#Orderings\n";
	for (auto i = plan.orderings.begin(); i != plan.orderings.end(); ++i) {
		out << i->first << " < " << i->second << '\n';
	}
	out << "\n#Causal Links\n";
	for (int i = 0; i < plan.links.size(); ++i) {
		out << plan.links[i].causalStep << ", ";
		printPredicate(out, plan.links[i].pred, tracker);
		out << ", " << plan.links[i].recipientStep << '\n';
	}
	out << "\n#Threats\n";
	for (auto i = plan.threats.begin(); i != plan.threats.end(); ++i) {
		out << i->actionId << ", (";
		out << i->threatened.causalStep << ", ";
		printPredicate(out, i->threatened.pred, tracker);
		out << ", " << i->threatened.recipientStep;
		out << ")\n";
	}
	out << "\n#Open Preconditions\n";
	for (int i = 0; i < plan.open.size(); ++i) {
		printPredicate(out, plan.open[i].first, tracker);
		out << '\n';
	}
}

File: read.py
'''
read.py
-------
Contains functions for reading and printing out plans
'''

from structures import *
from configure import *
from variables import *

def sAr(line): return filter(lambda x: len(x) > 0, line.strip().split())  ## Strip a line And Remove empty words

## Takes an empty plan object and an empty variable tracker
## Fills these with the info in the file specified by "filename"
def readfile(filename):
    infile = open(filename, "r")
    identifier = ''
    line = infile.readline()
    if not line:
        print "Could not read file."
        infile.close()
        return (False, 0, 0)

    locations = int(sAr(line)[-1])

    line = infile.readline()
    robots = int(sAr(line)[-1])

    line = infile.readline()
    cranes = int(sAr(line)[-1])

    line = infile.readline()
    piles = int(sAr(line)[-1])

    line = infile.readline()
    containers = int(sAr(line)[-1])

    tracker = VariableTracker(locations, robots, cranes, piles, containers)
    plan = Plan()
    plan.nextVar = tracker.getFirstVar()

    start = Action(Actions.START)
    end = Action(Actions.FINISH)
    goal = False

    line = infile.readline()
    while line:
        line = line.strip()
        if line.startswith('#') or len( line ) == 0:
            line = infile.readline()
            continue
        line_arr = sAr(line)
        identifier = line_arr[0]
        if (identifier == "initial"): 
            line = infile.readline()
            continue
        if (identifier == "goal"):
            goal = True
            line = infile.readline()
            continue

        try:
            p_id = Name2Predicate[identifier]
        except:
            print "Bad identifier"
            return (False, 0, 0)

        arg1 = arg2 = ''
        ## Two-argument propositions
        if p_id in [Predicates.ADJACENT, Predicates.ATTACHED, Predicates.BELONG, Predicates.AT, Predicates.LOADED, Predicates.HOLDING, Predicates.IN, Predicates.ON, Predicates.TOP]:
            try:
                arg1 = line_arr[1]
                arg2 = line_arr[2]
            except:
                print "Incomplete predicate"
                return (False, 0, 0)
            
            
            newPred = Predicate(p_id, tracker.getId(arg1), tracker.getId(arg2))

        ## One-argument propositions
        elif p_id in [Predicates.OCCUPIED, Predicates.UNLOADED, Predicates.EMPTY, Predicates.FREE]:
            try:
                arg1 = line_arr[1]
            except:
                print "Incomplete predicate"
                return (False, 0, 0)
            newPred = Predicate(p_id, tracker.getId(arg1))

        else: pass

        if (goal): plan.open_conditions.append((newPred,1))
        else: start.addList.append(newPred)
        
        line = infile.readline()

    plan.steps.append(start)
    plan.steps.append(end)

    plan.orderings.append((0,1))

    return (True, plan, tracker)



## Print a single predicate p
## Maybe useful for debugging
def printPredicate(p, tracker):
    print Predicate2Name[p.type_t],
    for i in range(2):
        if (p.args[i] < 0): break
        print tracker.getName(p.args[i]),


## Print a single action A
## Maybe useful for debugging
def printAction(A, tracker):
    print Action2Name[A.type_t],
    for i in range(5):
        if (A.args[i] < 0): break
        print tracker.getName(A.args[i]),

    print
    print "     Prereqs:",

    prereqs = A.getPrereqs()
    for i in range(len(prereqs)):
        printPredicate(prereqs[i], tracker)
        print ",",

    print
    print "     Added:",

    for i in range(len(A.addList)):
        printPredicate(A.addList[i], tracker)
        print ",",

    print
    print "     Deleted:",

    for i in range(len(A.deleteList)):
        printPredicate(A.deleteList[i], tracker)
        print ",",

    print


## A printing function that prints your plan.
## Use it for generating output files
def printPlan(plan, tracker):
    print "actions"
    startStep = endStep = -1
    for i in range(len(plan.steps)):
        act = plan.steps[i]
        if (act.type_t == Actions.START): startStep = i
        if (act.type_t == Actions.FINISH): endStep = i
        print i, str( act )

    print "\nconstraints"
    for i in range(len(plan.orderings)):
        o = plan.orderings[i]
        ## Skip over start / end ordering constraints
        if (o[0] == startStep or o[1] == endStep): continue
        print o[0], " < ", o[1]


## A function that prints more information about a plan
## Possibly useful for debugging
def printVerbosePlan(plan, tracker):
    print "#Steps"
    for i in range(len(plan.steps)):
        act = plan.steps[i]
        print i, Action2Name[act.type_t],
        for j in range(5):
            if (act.args[j] < 0): break
            print tracker.getName(act.args[j]),

        print

    print "\n#Orderings"
    for i in plan.orderings:
        print i[0], " < ", i[1]

    print "\n#Causal Links"
    for i in range(len(plan.links)):
        print plan.links[i].causalStep, ",",
        printPredicate(plan.links[i].pred, tracker)
        print ",", plan.links[i].recipientStep

    print "\n#Threats"
    for i in plan.threats:
        print i.actionId, ", (",
        print i.threatened.causalStep, ",",
        printPredicate(i.threatened.pred, tracker)
        print ",", i.threatened.recipientStep,
        print ")"

    print "\n#Open Preconditions"
    for i in range(len(plan.open)):
        printPredicate(plan.open[i][0], tracker)
        print


